sklearn/inspection/_ice.py

```diff
@@ -0,0 +1,491 @@
+"""Individual Conditional Expectation (ICE) plots for regression and
+classification models."""
+
+# Authors: Peter Prettenhofer
+#          Trevor Stephens
+#          Nicolas Hug
+#          Madhura Jayaratne
+# License: BSD 3 clause
+
+import numbers
+from itertools import count
+
+import numpy as np
+
+from ._base import (
+    _get_predictions, _validate_pdp_ice_parameters, _grid_from_X,
+    _plot_pdp_ice)
+from ..base import is_classifier
+from ..base import is_regressor
+from ..utils import _get_column_indices
+from ..utils import _safe_indexing
+from ..utils import check_matplotlib_support  # noqa
+
+__all__ = ['individual_conditional_expectation',
+           'plot_individual_conditional_expectation',
+           'IndividualConditionalExpectationDisplay']
+
+
+def _ice_brute(estimator, grid, features_indices, X, response_method):
+    predictions = _get_predictions(estimator, grid, features_indices, X,
+                                   response_method)
+    instances = X.shape[0]
+
+    # reshape to (n_targets, n_instances, n_points) where n_targets is:
+    # - 1 for non-multioutput regression and binary classification (shape is
+    #   already correct in those cases)
+    # - n_tasks for multi-output regression
+    # - n_classes for multiclass classification.
+    predictions = np.array(predictions).T
+    if is_regressor(estimator) and predictions.ndim == 2:
+        # non-multioutput regression, shape is (n_instances, n_points,)
+        predictions = predictions.reshape(instances, -1)
+    elif is_classifier(estimator) and predictions.shape[0] == 2:
+        # Binary classification, shape is (2, n_instances, n_points).
+        # we output the effect of **positive** class
+        predictions = predictions[1]
+        predictions = predictions.reshape(instances, -1)
+
+    return predictions
+
+
+def individual_conditional_expectation(estimator, X, features,
+                                       response_method='auto',
+                                       percentiles=(0.05, 0.95),
+                                       grid_resolution=100, centre=True,
+                                       **kwargs):
+    """Individual Conditional Expectation (ICE) of ``features``.
+
+    ICE of a feature (or a set of features) corresponds to the responses of an
+    estimator for each possible value of the feature for all instance in ``X``.
+
+    Read more in the :ref:`User Guide <individual_conditional_expectation>`.
```

Comment:
This user guide is not yet written, and will require you to update `doc/modules/partial_dependence.rst` perhaps renaming it

Reply:
Updated in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b04ba687c014b1ba9beaa63724574cc330d975c9. I feel 'Partial dependence' covers both PDPs and ICEs with PDP showing the average dependence while ICE visualizing the dependence for each instance separately.